PROJ-107 :: Added networking appendix (#54)
SW805F20/report: sections/appendix/network.tex

\section{Network packets}\label{app:network}

The final format of the network packages are described in this appendix.

> 

Reviewer: ```suggestion
```

> For all packages, it is assumed that the TCP messages are treated as unicast to a single recipient, and UDP are treated as multicast on socket.

Reviewer: ```suggestion
For all packages, it is assumed that the TCP messages are treated as unicast to a single recipient, and UDP messages are treated as multicast.
```

All of these packets will only describe the actual data of the packet, and not contain the UDP/TCP headers.

> 

Reviewer: ```suggestion
Each packet contains an ID identifying the type as well as the required information.
```

> \subsection*{0: Send tag position (UDP)}

Reviewer: ```suggestion

\subsection*{0: Send tag position (UDP)}
```

\begin{table}[H]
\begin{tabular}{|l|l|l|l|l|}
\hline

> \textbf{POSY (y)} & \textbf{POSX (x)} & \textbf{ID (i)} & \textbf{TIMESTAMP (s)} & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}TYPE (t)\\ 0\end{tabular}}} \\ \hline

Reviewer: overvejer lidt om vi skal have noget forklaring på størrelserne her 🤔

16 bit            & 16 bit            & 8 bit           & 8 bit                  & 8 bit                                                                              \\ \hline
\end{tabular}
\end{table}

Purpose: The position of a given tag will be continuously transmitted on the multicast channel for all players to receive.
The positions are sent as values relative to the anchor positions sent during setup, such that a missed package will not impact the future game state.

> The ID (i) \texttt{0} is reserved for the ball tag, and all other tags are considered to be players.

Reviewer: ```suggestion
The ID (i) \texttt{0} is reserved for the ball tag, and all other tags are considered to be players.
For a game with 4 players, each would have a unique ID ranging from 1-4.
```

\subsection*{1: Send anchor position (TCP)}
\begin{table}[H]
\begin{tabular}{|l|l|l|l|}
\hline
\textbf{POSY (y)} & \textbf{POSX (x)} & \textbf{ANCHOR\_ID (i)} & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}TYPE (t)\\ 1\end{tabular}}} \\ \hline
16 bit            & 16 bit            & 8 bit                   & 8 bit                                                                              \\ \hline
\end{tabular}
\end{table}

Purpose: Used as a part of the setup phase, where the host sends the position of one of the four anchors to the player.
This packet will be sent multiple times to each client, but with various anchor IDs (i) until all clients have received the four positions.

\subsection*{2: Send player tag (TCP)}
\begin{table}[H]
\begin{tabular}{|l|l|l|}
\hline

> \textbf{PLAYER\_ID (p)} & \textbf{TAG\_ID (i)} & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}TYPE (t)\\ 2\end{tabular}}} \\ \hline

Reviewer: hvorfor var det at tag  iderne er 16 bit?

This packet will be sent once to each client, but with various tag IDs (i) until all clients have received their tag ID.

\subsection*{3: Signal game start (TCP)}
\begin{table}[H]
\begin{tabular}{|l|}
\hline
\multicolumn{1}{|c|}{\textbf{\begin{tabular}[c]{@{}c@{}}TYPE (t)\\ 3\end{tabular}}} \\ \hline
8 bit                                                                               \\ \hline
\end{tabular}
\end{table}

Purpose: Used to signal to the players that the game has been started.

\subsection*{4: Send goal scored (TCP)}
\begin{table}[H]
\begin{tabular}{|l|l|l|}
\hline
\textbf{TEAM2SCORE (2)} & \textbf{TEAM1SCORE (1)} & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}TYPE (t)\\ 4\end{tabular}}} \\ \hline
8 bit                   & 8 bit                   & 8 bit                                                                              \\ \hline
\end{tabular}
\end{table}

Purpose: Used to signal the clients that a goal has been scored, and the scoreboard should be updated on their GUI.
This triggers the host to generate new goal positions.

> \subsection*{5: Send new goal position (TCP)}

Reviewer: det virker ve legentligt ikke det her ift til det der er skrevet, gør det?
sender den ikke midtpunkter for målene?
Så hvis vi sender en packet til hvert team vil de kun kende placeringen på en målzone, men ikke den anden 🤔

\begin{table}[H]
\begin{tabular}{|l|l|l|l|}
\hline
\textbf{POSY (y)} & \textbf{POSX (x)} & \textbf{TEAM\_ID (i)} & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}TYPE (t)\\ 5\end{tabular}}} \\ \hline
16 bit            & 16 bit            & 8 bit                 & 8 bit                                                                              \\ \hline
\end{tabular}
\end{table}

> Purpose: Sends a new goal position to the clients when a goal is scored. The packet will be sent twice, one packet for each team with different team IDs (i).

Reviewer: ```suggestion
Purpose: Sends a new goal position to the clients when a goal is scored.
The packet will be sent twice, one packet for each team with different team IDs (i).
```

\end{tabular}
\end{table}

Purpose: Used to signal to the players that the game has been started.

\subsection*{4: Send goal scored (TCP)}
\begin{table}[H]
\begin{tabular}{|l|l|l|}
\hline
\textbf{TEAM2SCORE (2)} & \textbf{TEAM1SCORE (1)} & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}TYPE (t)\\ 4\end{tabular}}} \\ \hline
8 bit                   & 8 bit                   & 8 bit                                                                              \\ \hline
\end{tabular}
\end{table}

Purpose: Used to signal the clients that a goal has been scored, and the scoreboard should be updated on their GUI.
This triggers the host to generate new goal positions.

\subsection*{5: Send new goal position (TCP)}
\begin{table}[H]
\begin{tabular}{|l|l|l|l|}
\hline
\textbf{POSY (y)} & \textbf{POSX (x)} & \textbf{TEAM\_ID (i)} & \multicolumn{1}{c|}{\textbf{\begin{tabular}[c]{@{}c@{}}TYPE (t)\\ 5\end{tabular}}} \\ \hline
16 bit            & 16 bit            & 8 bit                 & 8 bit                                                                              \\ \hline
\end{tabular}
\end{table}

Purpose: Sends a new goal position to the clients when a goal is scored.

> The packet will be sent twice, one packet for each team with different team IDs (i).

Author: ```suggestion
The packet will be sent twice to each team, once per goal.
The positions are the X and Y coordinates of the center of the goal as they have a pre-defined size based on the shortest edge of the playing field.```
